File: Node/merge-sort.js
/**
 *  Author: CJ Galindo
 *  Merge sort algorigthm
 *  Time complexity: O (n log n)
 *  Best time: Ω (n log n)
 *  Space Complexity: O(n)
 */
function mergeSort (arr) {
  const len = arr.length

  if (len === 1) {
    return arr
  }

  const mid = Math.floor(len / 2)
  const left = mergeSort(arr.slice(0, mid))
  const right = mergeSort(arr.slice(mid))

  return merge(left, right)
}

function merge (left, right) {
  let arr = []
  const lenLeft = left.length
  const lenRight = right.length
  let indexLeft = 0
  let indexRight = 0
  while (indexLeft < lenLeft && indexRight < lenRight) {
    if (left[indexLeft] < right[indexRight]) {
      arr.push(left[indexLeft])
      indexLeft++
    } else {
      arr.push(right[indexRight])
      indexRight++
    }
  }

  while (indexLeft < lenLeft) {
    arr.push(left[indexLeft])
    indexLeft++
  }

  while (indexRight < lenRight) {
    arr.push(right[indexRight])
    indexRight++
  }

  return arr
}


console.log(mergeSort([5,4,3,2,1]))

File: Python/merge-sort.py
import math

def merge_sort(arr):
    count = len(arr)

    if count == 1:
        return arr

    mid = math.floor(count / 2)

    left = merge_sort(arr[0:mid])
    right = merge_sort(arr[mid:])

    return merge(left, right)

def merge(left, right):
    arr = []
    len_left = len(left)
    len_right = len(right)
    index_left, index_right = 0,0

    while index_left < len_left and index_right < len_right:
        if left[index_left] < right[index_right]:
            arr.append(left[index_left])
            index_left += 1
        else:
            arr.append(right[index_right])
            index_right +=1
    
    while index_left < len_left:
        arr.append(left[index_left])
        index_left += 1

    while index_right < len_right:
        arr.append(right[index_right])
        index_right += 1
    
    return arr

print(merge_sort([9,8,7,5,3,6,2,1,4]))